api/v1/relay/relay.proto

```diff
@@ -0,0 +1,152 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+import "google/protobuf/timestamp.proto";
+import public "flow/flow.proto";
+
+package relay;
+
+service HubbleRelay {
+    rpc GetEvents(GetEventsRequest) returns (stream GetEventsResponse) {}
+}
+
+// Here I didn't include "follow", "until", and "number". This request assumes follow,
+// and lets the client decide when to end the request, whether it's based on timestamp
+// or the number of responses received.
+message GetEventsRequest{
+    // EventType specifies which types of events to subscribe to. If unspecified, the
+    // request is subscribed to all the event types.
+    repeated EventType event_types = 1;
+    repeated EventFilter blacklist = 2;
+    repeated EventFilter whitelist = 3;
+    google.protobuf.Timestamp since = 4;
+}
+
+message GetEventsResponse{
+    string node = 1;
+    google.protobuf.Timestamp timestamp = 2;
+    oneof event {
+        flow.Flow flow = 3;
+        K8sNamespaceState k8s_namespace_state = 4;
+        ServiceState service_state = 5;
+        // It's unclear how we can define what it means for service links to get
+        // added/deleted. For now I'm deferring the decision for whether a particular
+        // link should be included in the service map to the client. Maybe the server
+        // can periodically send service_link events for active links to keep-alive,
+        // and the client can expire older links based on their timestamp.
+        ServiceLinkState  service_link_state = 6;
+    }
+}
+
+// Here I added different filter types for different events. Another option is to reuse
+// and extend FlowFilter to filter all the events, but I worry that it might become too
+// confusing in terms of figuring out which filter fields apply to which events.
+message EventFilter {
+    oneof filter {
+        flow.FlowFilter flow_filter = 2;
+        ServiceFilter service_filter = 3;
+        ServiceLinkFilter service_link_filter = 4;
+    }
+}
+
+enum EventType {
+    UNKNOWN_EVENT = 0;
+    FLOW = 1;
+    K8S_NAMESPACE_STATE = 2;
+    SERVICE_STATE = 3;
+    SERVICE_LINK_STATE = 4;
+}
+
+message K8sNamespace {
+    string id = 1;
+    string name = 2;
+    google.protobuf.Timestamp creation_timestamp = 3;
+}
+
+message K8sNamespaceState {
+    K8sNamespace namespace = 1;
+    StateChange type = 2;
+}
```

Comment:
I think that these 2 should be grouped together. Same for Service/ServiceState and ServiceLink/ServiceLinkState.